File: 프로그래머스/KaKao/개인정보 수집 유효기간.swift
// 2023 KAKAO BLIND RECRUITMENT 개인정보 수집 유효기간
// https://school.programmers.co.kr/learn/courses/30/lessons/150370
import Foundation

func solution(_ today:String, _ terms:[String], _ privacies:[String]) -> [Int] {
    var answer: [Int] = []
    let intToday = getIntegerTotalDay(today)
    let termsDic = translateTermsForDictionay(terms)
    
    for i in 0..<privacies.count {
        let splitedPrivacy = privacies[i].components(separatedBy: " ")
        let currentInfo = getIntegerTotalDay(splitedPrivacy[0])
        let termType = termsDic[splitedPrivacy[1]]!
        
        if intToday >= currentInfo + termType {
            answer.append(i+1)
        }
    }
    return answer
}

func getIntegerTotalDay(_ day: String) -> Int {
    let splitedDay = day.components(separatedBy: ".")
    return Int(splitedDay[0])! * 12 * 28 + Int(splitedDay[1])! * 28 + Int(splitedDay[2])!
}

func translateTermsForDictionay(_ terms:[String]) -> [String:Int] {
    var ret:[String:Int] = [:]
    
    for term in terms {
        let splitedTerms = term.components(separatedBy: " ")
        ret[splitedTerms[0]] = Int(splitedTerms[1])! * 28 
    }
    
    return ret
}

File: 프로그래머스/KaKao/개인정보 수집 유효기간.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/150370
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

int vectorString_to_Int(vector<string> v) {
    int ret = (stoi(v[0])*12*28)+(stoi(v[1])*28)+stoi(v[2]);
    return ret;
}

vector<string> split(string s, string ch) {
    vector<string> ret;
    
    while (s.find(ch) != string::npos) 
    {
        int idx = s.find(ch);
        ret.push_back(s.substr(0, idx));
        s = s.substr(idx+ch.length());
    }
    
    ret.push_back(s);
    
    return ret;
}

vector<int> solution(string today, vector<string> terms, vector<string> privacies) {
    vector<int> answer;
    unordered_map<string,int> termsMap;
    int Today = vectorString_to_Int(split(today, "."));
    
    for (string& t: terms)
    {
        vector<string> splitedT = split(t, " ");
        termsMap[splitedT[0]] = stoi(splitedT[1]) * 28;
    }
    
    for (int i = 0; i < privacies.size(); ++i)
    {
        vector<string> tmpPri = split(privacies[i], " ");
        int currentDay = vectorString_to_Int(split(tmpPri[0], "."));
        int currentT = termsMap[tmpPri[1]];
        
        //cout << currentDay+currentT << ' ' << Today << '\n';
        if (currentDay + currentT <= Today) {
            answer.push_back(i+1);
        }
    }
    
    return answer;
}